Update Cmsis-pack-manager to 0.2.3

theotherjimmy opened the pull request:
### Description

This comes with quite a few features:
 * Variant parsing
 * Memory permission bits for:
   * Read
   * Write
   * Execute
   * Secure
   * Non-Secure
   * Non-Secure Callable
   * Peripheral
 * Memory default information
 * Descriptions of core as either:
   * Semetric Multi-core
   * Asymetric Multi-core

Further, The index has been re-formatted to be human-readable and all 
keys are sorted so that diffs are also human-readable. Be aware, it's 
big.

~NOTE: This will conflict with #8607~ Rebased after #8607


### Pull request type

    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

theotherjimmy committed:
Start using CPM

theotherjimmy committed:
Use upgrade table for unavailable CMP data

theotherjimmy committed:
Use updated CPM info in config system

cmonr commented:
_Looks at changes_
_Sees 450k+ changes_
Uhhh....

_Sees_ tools/arm_pack_manager/index.json _was updated_
Oh. Carry on.


@theotherjimmy Would/could that file benefit from getting new lines added to it, similar to how target.json was updated, so that the diff generated isn't 99.9% if the PR when it's updated?

cmonr commented on requirements.txt:
```diff
@@ -19,3 +19,4 @@ six==1.11.0
 git+https://github.com/armmbed/manifest-tool.git@v1.4.6
 mbed-cloud-sdk==2.0.1
 icetea>=1.0.2,<1.1
+cmsis-pack-manager>=0.2.2,<0.3.0
```
🎉

bridadan commented on tools/config/__init__.py:
```diff
@@ -600,15 +600,26 @@ def _get_cmsis_part(self):
+    def _get_mem_specs(self, permissions, cmsis_part):
+        all_matching_memories = [
+            memory for memory in cmsis_part['memories'].values()
+            if all(memory['access'][perm] for perm in permissions)
```
Is it guarenteed that the key `perm` will be in the dictionary `memory['access']`? If not, this would be a safer alternative (no risk of `KeyError`):
```suggestion
            if all(memory['access'].get(perm) for perm in permissions)
```

theotherjimmy replied:
Technically it's guaranteed. OTOH, it's safer to do it the way you recommend.

cmonr commented on tools/arm_pack_manager/__init__.py:
```diff
@@ -1,309 +1,104 @@
         """Extract sector sizes from device FLM algorithm
 
-        Will return None if there is no algorithm, pdsc URL formatted in correctly
+        Will return None if there is no algorithm, pdsc URL formatted in
```
Not sure why, but this feels like weird english.

Is it missing an `or`?

theotherjimmy replied:
Yes.

bridadan commented on tools/config/__init__.py:
```diff
@@ -600,15 +600,26 @@ def _get_cmsis_part(self):
+        ]
+        if all_matching_memories:
+            memory = sorted(
+                all_matching_memories, key=self._memory_ordering
```
I'm a bit tripped up on the `key` function here. Is this essentially doing the following for the comparison:

"
For each `memory` in `all_matching_memories`, the one that is the "least" is the one that:

1. Has the smallest value for `memory['default']`
2. If they are the same, pick the smallest value for `memory['size']`
3. If that is the same, pick the smallest value for `memory['start']`

"

theotherjimmy replied:
@bridadan I was under the impression that it would be largest first. If I need a quick reverse to make them all "largest" then that's what I'll do.

bridadan replied:
`sorted` will return an ascending order sorted list by default: https://docs.python.org/3/howto/sorting.html. So that would mean the first one in the list would be the "smallest" if I'm interpreting the doc correctly.

theotherjimmy replied:
Right. Adding a reverse now.

cmonr commented on tools/arm_pack_manager/pack_manager.py:
```diff
@@ -1,192 +0,0 @@
-from __future__ import print_function, division, absolute_import
```
_Goodnight sweet prince_

bridadan reviewed (commented):
Few clarifications

cmonr reviewed (approved):
Nice!

cmonr commented:
@ARMmbed/mbed-os-test-team Heads up. A new python module is being added into mbed-os

https://github.com/ARMmbed/mbed-os/pull/8757/files#diff-b4ef698db8ca845e5845c4618278f29aR22


@theotherjimmy Do you know if `cmsis-pack-manager` has a minimum Py2/Py3 versions?

bridadan committed:
Avoid KeyError Exceptions.

Co-Authored-By: theotherjimmy <theotherjimmy@gmail.com>

theotherjimmy commented:
@cmonr It's compatible with most Python versions >= 2.7

theotherjimmy commented:
@cmonr 
> Would/could that file benefit from getting new lines added to it, similar to how target.json was updated, so that the diff generated isn't 99.9% if the PR when it's updated?

from the top comment:

>Further, The index has been re-formatted to be human-readable and all
keys are sorted so that diffs are also human-readable. Be aware, it's
big.

cmonr commented:
> Further, The index has been re-formatted to be human-readable and all
keys are sorted so that diffs are also human-readable. Be aware, it's
big.

Ha. Silly GitHub not wanting to load the file...

NirSonnenschein commented:
/morph build

NirSonnenschein commented:
@0xc0170 , @cmonr  I've advanced this one, any reason why it still needs work?

NirSonnenschein commented:
@theotherjimmy  this PR ran into a strange build failure, mind taking a look:
 Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mbed/mbed.py", line 3338, in main
    status = pargs.command(pargs)
  File "/usr/local/lib/python2.7/dist-packages/mbed/mbed.py", line 1985, in thunk
    return command(**argv)
  File "/usr/local/lib/python2.7/dist-packages/mbed/mbed.py", line 2769, in test_
    icetea_supported = program.requirements_contains('icetea')
  File "/usr/local/lib/python2.7/dist-packages/mbed/mbed.py", line 1603, in requirements_contains
    with open(os.path.join(req_path, req_file), 'r') as f:
IOError: [Errno 2] No such file or directory: '/builds/ws/mbed-os-build-matrix/3666/default_GCC_ARM/requirements.txt'

0xc0170 commented:
> NOTE: The history is not clean yet. I plan to squash/rewrite commit
messages to meet commit message standards.

@NirSonnenschein labeled as needs: work because of this

OPpuolitaival commented on tools/arm_pack_manager/__init__.py:
```diff
@@ -1,309 +1,104 @@
-                    flm = pack.open(filename)
+                    flm = pack.open(
+                        algo["file_name"]
+                        .replace("\\\\", "/")
```
This works also in Windows?

theotherjimmy replied:
Maybe. I'm not sure.

OPpuolitaival reviewed (commented):
Is this tested also in Windows machine?

theotherjimmy commented:
@OPpuolitaival This has not been tested on a windows machine that's next on my list.

theotherjimmy commented:
@0xc0170 The history is actually not so bad. I already did some squashing/rewriting.

deepikabhavnani commented:
@theotherjimmy - as discussed internally should we move this to 5.12 as it will have broader impact on devices?

theotherjimmy commented:
@deepikabhavnani Perhaps "broader impact" is not the correct phrase. "it's a big risk" as it can impact many systems is probably what you were thinking. Especially because it's supposed to have no impact on any devices.

deepikabhavnani commented:
@theotherjimmy - Yes agree risk is higher and we might uncover some unknowns as well..

0xc0170 commented:
Based on the recent discussion, this was moved to 5.12 release version

cmonr commented:
Sure enough, now that #8607 is in, this needs a rebase.

SenRamakri commented:
Why is this marked for 5.12 vs 5.11.1? @aashishc1988 has been helping with reproducing https://github.com/ARMmbed/mbed-os/issues/5467 and found this change to be fixing the problem. What is needed to merge this earlier?

aashishc1988 commented:
@SenRamakri I think there was some miscommunication. My statement might have been misleading. This PR should ideally fix the issue, but I havent been able to test it yet because it needs little more work.

SenRamakri commented:
@aashishc1988 - Thanks for clarifying, please update this thread once you have tested the scenario with this fix.

theotherjimmy committed:
Correct uvisionn export

theotherjimmy committed:
Use device variant in uvision export

theotherjimmy committed:
Update index to include vendor information

0xc0170 commented:
CI started

0xc0170 commented:
Failed early,

```
[examples] Running shell script

+ python -u ../mbed-os/tools/test/examples/examples.py clone

Traceback (most recent call last):

  File "../mbed-os/tools/test/examples/examples.py", line 16, in <module>

    from tools.build_api import get_mbed_official_release

  File "/builds/ws/ci_PR-8757-C76PPJ3GVWAIAMWKIPE3J/mbed-os/tools/build_api.py", line 35, in <module>

    from .arm_pack_manager import Cache

  File "/builds/ws/ci_PR-8757-C76PPJ3GVWAIAMWKIPE3J/mbed-os/tools/arm_pack_manager/__init__.py", line 4, in <module>

    from cmsis_pack_manager import Cache as _Cache

ImportError: No module named cmsis_pack_manager

script returned exit code 1
```

Does this require any CI update or just a bug in tools?

0xc0170 commented:
@theotherjimmy Can you create internal ticket for MBEDOSTEST and assign @OPpuolitaival .

cmonr commented:
@0xc0170 I also don't think that the PR was ready for testing.

theotherjimmy commented:
Status: I think the uvision + STM "no device found" bugs are resolved by this PR, and It required a modification to CPM, so I'll need to update the version of CPM used.

cmonr commented:
Waiting on https://github.com/ARMmbed/mbed-os/pull/9056 to come in.

cmonr commented:
@theotherjimmy It's ready to progress!

theotherjimmy commented:
@cmonr Good theory. I need to release 0.2.3 to add vendor parsing to CMP (already present in the index generated by this PR) so that when someone runs this again, we don't loose all of that information.

aashishc1988 commented:
@SenRamakri

theotherjimmy commented:
Alright. I need to wait for assets [here](https://github.com/ARMmbed/cmsis-pack-manager/releases/tag/v0.2.3) and release to pip.

theotherjimmy committed:
Require 0.2.3 CPM

This includes vendor parsing so that we can reproduced the index.json

theotherjimmy committed:
Add old name for STM32H743ZI

theotherjimmy commented:
@cmonr @0xc0170 CPM 0.2.3 is released and installable through pip. This should no longer need work.

cmonr commented:
CI started

cmonr commented:
Looks like the module isn't being brought into CI.

Ping'ed @ARMmbed/mbed-os-test-team

0xc0170 commented:
We need to raise an internal ticket for updating CI packages

cmonr commented:
Ticket raised: https://jira.arm.com/browse/MBEDOSTEST-464

theotherjimmy committed:
Correct name of K22F

theotherjimmy committed:
Recompute sector information for STM32F429ZITx

theotherjimmy committed:
Correct python3 bug in gerenate index

theotherjimmy committed:
Recreate STM32F439ZITx sector information

theotherjimmy committed:
Recompute Numaker PFM M2351 sectors

theotherjimmy committed:
Manually re-combine NUMAKER_PFM_M2351 memories

0xc0170 committed:
nuvoton m2351: fix RAM regions - only one RAM1

Revert latest change to index. Linker scripts follow one RAM region. If index is updated,
requires further changes in the target that should be done separately.

theotherjimmy committed:
Remove prints and lint warnings
